shell/platform/fuchsia/flutter/vsync_waiter.cc

```diff
@@ -61,14 +64,56 @@ VsyncWaiter::~VsyncWaiter() {
   ui_latch.Wait();
 }
 
-static fml::TimePoint SnapToNextPhase(fml::TimePoint value,
-                                      fml::TimePoint phase,
-                                      fml::TimeDelta interval) {
-  fml::TimeDelta offset = (phase - value) % interval;
-  if (offset < fml::TimeDelta::Zero()) {
-    offset = offset + interval;
+/// Returns the system time at which the next frame is likely to be presented.
+///
+/// Consider the following timelines, where in both the scenarious the result
+/// will be the same.
+///
+/// Scenario 1:
+/// presentation_interval is 2
+/// ^     ^     ^     ^     ^
+/// +     +     +     +     +
+/// 0--1--2--3--4--5--6--7--8--9--
+/// +        +  +
+/// |        |  +---------> result: next_presentation_time
+/// |        v
+/// v        now
+/// last_presentation_time
+///
+/// Scenario 2:
+/// presentation interVal is 2
+/// ^     ^     ^     ^     ^
+/// +     +     +     +     +
+/// 0--1--2--3--4--5--6--7--8--9--
+///       +  +  +
+///       |  |  +--------->result: next_presentation_time
+///       |  |
+///       |  +>now
+///       |
+///       +->last_presentation_time
+fml::TimePoint VsyncWaiter::SnapToNextPhase(
+    const fml::TimePoint now,
+    const fml::TimePoint last_frame_presentation_time,
+    const fml::TimeDelta presentation_interval) {
+  FML_CHECK(presentation_interval > fml::TimeDelta::Zero())
+      << "Presentation interval whould be positive.";
+  FML_DCHECK(last_frame_presentation_time < now)
+      << "Last frame was presented in the future. This is unexpected.";
+
+  const fml::TimeDelta delta = now - last_frame_presentation_time;
+  // this will be the most likely scenario if we are rendering at a good
+  // frame rate, short circuiting the other checks in this case.
+  if (delta < presentation_interval) {
+    return last_frame_presentation_time + presentation_interval;
+  } else {
+    int64_t num_phases_passed = delta / presentation_interval;
+    // this is to ensure that we will be past the "now" time point.
+    if (delta % presentation_interval == fml::TimeDelta::Zero()) {
```

Comment:
I'm not sure I fully understand this. I think it's supposed to be iterating over to the first future timepoint at which a presentation is scheduled to occur, so e.g. if we have a delta of 20ms since the last presentation, and the presentation interval is 16ms, we're 4ms into the budget for the n+1 frame, which means we need to return a timepoint 12ms in the future from now, is this correct?

In this expression, 20 % 16 != 0 so we won't be adding to the phases passed, and the integer division above should return the floor so we'd be returning a timepoint in the past?

`last_frame_presentation_time` = `0`
`presentation_interval` = `16`
`now` = `20`
`delta` = `20`
`num_phases_passed` = `20 / 16` = `1`
`delta % presentation_interval` = `20 % 16` = `4`

return `0 + (16 * 1)` = `16`?

Reply:
I meant to do a `ceil` and was accounting for the case where there is a perfect division. good catch, will add a test for this.